File: src/pharmgmt/static/js/pages/reports.js
/* Reports Page */
async function renderReports(container) {
    container.innerHTML = `
    <div class="page-header"><h2>Reports</h2><p>Purchase reports, stock summaries, and sanity checks</p></div>

    <div class="card mb-6">
      <div class="flex justify-between items-center mb-4">
        <h3>📦 Purchase Report</h3>
        <div class="flex gap-4">
          <a href="/api/reports/purchases/csv" class="btn btn-secondary btn-sm" target="_blank">📥 CSV</a>
          <button class="btn btn-ghost btn-sm" onclick="window.print()">🖨 Print</button>
        </div>
      </div>
      <div id="purchase-report"><div class="skeleton skeleton-card" style="height:150px"></div></div>
    </div>

    <div class="card mb-6">
      <div class="flex justify-between items-center mb-4">
        <h3>📊 Stock Summary</h3>
        <div class="flex gap-4">
          <a href="/api/reports/stock/csv" class="btn btn-secondary btn-sm" target="_blank">📥 CSV</a>
        </div>
      </div>
      <div id="stock-report"><div class="skeleton skeleton-card" style="height:150px"></div></div>
    </div>

    <div class="card mb-6">
      <h3 class="mb-4">🔍 Sanity Report</h3>
      <div id="sanity-report"><div class="skeleton skeleton-card" style="height:150px"></div></div>
    </div>`;

    // Purchase Report
    try {
        const pr = await API.request('/api/reports/purchases');
        if ((pr.items || []).length === 0) {
            document.getElementById('purchase-report').innerHTML = `<p style="color:var(--text-secondary)">No purchase data yet</p>`;
        } else {
            document.getElementById('purchase-report').innerHTML = `
        <div class="grid-2 mb-4">
          <div class="stat-card"><div class="stat-value">${pr.total_items}</div><div class="stat-label">Items</div></div>
          <div class="stat-card"><div class="stat-value">${formatMoney(pr.total_value_paise)}</div><div class="stat-label">Total Value</div></div>
        </div>
        <div class="table-container" style="max-height:300px;overflow-y:auto"><table class="data-table"><thead><tr>
          <th>File</th><th>Product</th><th>Pack</th><th>Qty</th><th>Price</th><th>Value</th>
        </tr></thead><tbody>${pr.items.slice(0, 50).map(i => `<tr>
          <td>${i.file_name || '—'}</td><td>${i.product || '—'}</td><td>${i.packing || '—'}</td>
          <td>${i.quantity || 0}</td><td>${formatMoney(i.price_paise)}</td><td>${formatMoney(i.value_paise)}</td>
        </tr>`).join('')}</tbody></table></div>`;
        }
    } catch (e) { document.getElementById('purchase-report').innerHTML = `<p style="color:var(--danger)">Error: ${e.message}</p>`; }

    // Stock Summary
    try {
        const st = await API.request('/api/reports/stock');
        if ((st.items || []).length === 0) {
            document.getElementById('stock-report').innerHTML = `<p style="color:var(--text-secondary)">No stock data yet</p>`;
        } else {
            document.getElementById('stock-report').innerHTML = `
        <div class="grid-2 mb-4">
          <div class="stat-card"><div class="stat-value">${st.total_products}</div><div class="stat-label">Products</div></div>
          <div class="stat-card"><div class="stat-value">${formatMoney(st.total_value_paise)}</div><div class="stat-label">Total Value</div></div>
        </div>
        <div class="table-container" style="max-height:300px;overflow-y:auto"><table class="data-table"><thead><tr>
          <th>Product</th><th>Pack</th><th>Stock</th><th>Price</th><th>Value</th><th>Expiry</th>
        </tr></thead><tbody>${st.items.map(i => `<tr>
          <td>${i.product || '—'}</td><td>${i.packing || '—'}</td>
          <td>${i.closing_qty || 0}</td><td>${formatMoney(i.price_paise)}</td>
          <td>${formatMoney(i.value_paise)}</td><td>${i.expiry || '—'}</td>
        </tr>`).join('')}</tbody></table></div>`;
        }
    } catch (e) { document.getElementById('stock-report').innerHTML = `<p style="color:var(--danger)">Error: ${e.message}</p>`; }

    // Sanity Report
    try {
        const san = await API.request('/api/reports/sanity');
        if ((san.flagged_documents || []).length === 0) {
            document.getElementById('sanity-report').innerHTML = `<div class="flex items-center gap-4"><span style="color:var(--success);font-size:1.5rem">✅</span><span>No issues found — all documents parsed cleanly</span></div>`;
        } else {
            document.getElementById('sanity-report').innerHTML = `
        <div class="grid-2 mb-4">
          <div class="stat-card"><div class="stat-value">${san.total_flagged_docs}</div><div class="stat-label">Flagged Docs</div></div>
          <div class="stat-card"><div class="stat-value">${san.total_flagged_rows}</div><div class="stat-label">Flagged Rows</div></div>
        </div>
        <div class="table-container"><table class="data-table"><thead><tr><th>File</th><th>Confidence</th><th>Flagged Rows</th><th>Issues</th><th></th></tr></thead><tbody>
        ${san.flagged_documents.map(d => `<tr>
          <td>${d.file_name || '—'}</td>
          <td>${confidenceBar(d.avg_confidence)}</td>
          <td>${d.rows_flagged}</td>
          <td>${(d.error_flags || []).map(f => `<span class="badge badge-warning">${f}</span>`).join(' ')}</td>
          <td><a href="#/bills/${d.document_id}" class="btn btn-ghost btn-sm">View →</a></td>
        </tr>`).join('')}</tbody></table></div>`;
        }
    } catch (e) { document.getElementById('sanity-report').innerHTML = `<p style="color:var(--danger)">Error: ${e.message}</p>`; }
}

File: src/pharmgmt/services/reports.py
"""Reports service — purchase reports, stock summaries, sanity reports."""

import json
import logging

logger = logging.getLogger("pharmgmt.reports")


def purchase_report(session, date_from: str = None, date_to: str = None) -> dict:
    """Generate purchase report for a date range.

    Groups line items by document (supplier), then product.

    Returns:
        Dict with items list and totals
    """
    from pharmgmt.models import LineItem, Document

    query = session.query(LineItem, Document).join(Document, LineItem.document_id == Document.id)

    if date_from:
        query = query.filter(Document.ingest_ts >= date_from)
    if date_to:
        query = query.filter(Document.ingest_ts <= date_to)

    results = query.order_by(Document.ingest_ts.desc()).all()

    items = []
    total_qty = 0
    total_value = 0

    for li, doc in results:
        qty = li.issue_qty or li.closing_qty or 0
        price = li.price_paise or 0
        value = qty * price

        items.append({
            "document_id": doc.id,
            "file_name": doc.file_name,
            "date": doc.ingest_ts.isoformat() if doc.ingest_ts else None,
            "product": li.product_name_raw,
            "packing": li.packing,
            "batch": li.batch_no,
            "quantity": qty,
            "price_paise": price,
            "value_paise": value,
        })
        total_qty += qty if isinstance(qty, (int, float)) else 0
        total_value += value if isinstance(value, (int, float)) else 0

    return {
        "items": items,
        "total_items": len(items),
        "total_quantity": total_qty,
        "total_value_paise": total_value,
    }


def stock_summary(session) -> dict:
    """Generate current stock snapshot.

    Per product: latest closing qty, latest price, total value.

    Returns:
        Dict with products list and totals
    """
    from sqlalchemy import func
    from pharmgmt.models import LineItem

    products = (
        session.query(
            LineItem.product_name_raw,
            LineItem.packing,
            func.max(LineItem.closing_qty).label("closing"),
            func.max(LineItem.price_paise).label("price"),
            func.max(LineItem.expiry).label("expiry"),
        )
        .filter(LineItem.product_name_raw.isnot(None))
        .group_by(LineItem.product_name_raw, LineItem.packing)
        .order_by(LineItem.product_name_raw)
        .all()
    )

    items = []
    total_value = 0
    for p in products:
        closing = p.closing or 0
        price = p.price or 0
        value = closing * price if isinstance(closing, (int, float)) and isinstance(price, (int, float)) else 0
        items.append({
            "product": p.product_name_raw,
            "packing": p.packing,
            "closing_qty": closing,
            "price_paise": price,
            "value_paise": value,
            "expiry": p.expiry,
        })
        total_value += value

    return {"items": items, "total_products": len(items), "total_value_paise": total_value}


def sanity_report(session) -> dict:
    """Generate sanity report with all flagged issues.

    Returns:
        Dict with flagged documents, rows, and summary
    """
    from pharmgmt.models import ParsingRun, Document

    runs = session.query(ParsingRun).all()

    flagged_docs = []
    total_flagged_rows = 0

    for pr in runs:
        flags = json.loads(pr.error_flags) if pr.error_flags else []
        if not flags and pr.rows_flagged == 0:
            continue

        doc = session.query(Document).filter_by(id=pr.document_id).first()
        flagged_docs.append({
            "document_id": pr.document_id,
            "file_name": doc.file_name if doc else None,
            "avg_confidence": pr.avg_confidence,
            "rows_flagged": pr.rows_flagged,
            "error_flags": flags,
            "needs_review": bool(pr.needs_review),
        })
        total_flagged_rows += pr.rows_flagged or 0

    return {
        "flagged_documents": flagged_docs,
        "total_flagged_docs": len(flagged_docs),
        "total_flagged_rows": total_flagged_rows,
    }
